File: backend/14-notification/src/db/redis.ts
// Redis DB 10 — dedup window + WebSocket pod routing + cross-pod pub/sub fan-out.
// DEGRADABLE: Redis is NEVER a /ready gate (architecture §8.1) and a connection
// error must NOT crash the process. We lazyConnect, swallow 'error' events to a
// warning, and the dedup path FAILS OPEN (treats Redis-down as "not a duplicate")
// so an outage degrades to possible-redelivery rather than dropped notifications.
import Redis from 'ioredis';
import { config } from '../config';
import { logger } from '../observability/logger';

// Cross-pod WS broadcast travels over ONE Redis channel; each pod filters by userId.
export const WS_CHANNEL = 'ws:user';

// Payload shape on the WS_CHANNEL pub/sub (documented for the realtime agent):
//   { userId: string; payload: any }
export interface WsBroadcast {
  userId: string;
  payload: any;
}

let client: Redis | null = null;
let subscriber: Redis | null = null;

function build(): Redis {
  const r = new Redis({
    host: config.redisHost,
    port: config.redisPort,
    password: config.redisPassword || undefined,
    db: config.redisDb,
    lazyConnect: true,
    maxRetriesPerRequest: 2,
    enableReadyCheck: true,
    retryStrategy: (times) => Math.min(times * 200, 2000),
  });
  // Swallow errors to a warning — Redis must never crash the process.
  r.on('error', (e: any) => logger.warn('notification.redis', `redis error: ${String(e?.message).slice(0, 80)}`));
  return r;
}

// Best-effort connect (non-fatal). bootstrap() does NOT await readiness as a gate.
export async function connectRedis(): Promise<void> {
  if (client) return;
  client = build();
  try {
    await client.connect();
    logger.info('notification.boot', 'redis connected (db 10)');
  } catch (e: any) {
    logger.warn('notification.boot', `redis connect deferred: ${String(e?.message).slice(0, 80)}`);
  }
}

export function getRedis(): Redis {
  if (!client) throw new Error('redis not initialised');
  return client;
}

// Liveness check for /health ONLY (diagnostic — never the /ready gate). Throws on failure.
export async function ping(): Promise<void> {
  if (!client) throw new Error('redis not initialised');
  await client.ping();
}

// Dedup: SET notif:dedup:<topic>:<key> '1' NX EX ttl.
//   returns true  → newly set (this is the FIRST delivery → process it)
//   returns false → key already present (CONFIRMED duplicate → skip)
// FAILS OPEN: on any Redis error returns true (process the message) so a Redis
// outage degrades to possible-redelivery, never to a dropped notification.
export async function dedupSetNx(topic: string, key: string, ttl = config.notifDedupTtl): Promise<boolean> {
  if (!client) return true;
  try {
    const res = await client.set(`notif:dedup:${topic}:${key}`, '1', 'EX', ttl, 'NX');
    return res === 'OK';
  } catch (e: any) {
    logger.warn('notification.redis', `dedup degraded (fail-open): ${String(e?.message).slice(0, 60)}`);
    return true;
  }
}

// WS routing: map a connected user -> the pod id holding their socket (session TTL).
export async function setWsUserPod(userId: string, podId: string, ttlSeconds = 3600): Promise<void> {
  if (!client) return;
  try { await client.set(`ws:user:${userId}`, podId, 'EX', ttlSeconds); } catch { /* degrade */ }
}
export async function getWsUserPod(userId: string): Promise<string | null> {
  if (!client) return null;
  try { return await client.get(`ws:user:${userId}`); } catch { return null; }
}

// Cross-pod fan-out: publish a WS event so the pod holding the user's socket (which
// may be a different pod than the consumer) can push it. Best-effort.
export async function publishWsEvent(userId: string, payload: any): Promise<void> {
  if (!client) return;
  try { await client.publish(WS_CHANNEL, JSON.stringify({ userId, payload } as WsBroadcast)); } catch { /* degrade */ }
}

// Subscribe to the cross-pod WS channel. Uses a DEDICATED duplicated connection —
// a subscribed ioredis client cannot run normal commands. handler receives the
// parsed { userId, payload }; the realtime layer filters to its locally-held sockets.
export async function subscribeWsEvents(handler: (msg: WsBroadcast) => void): Promise<void> {
  if (!client) return;
  if (!subscriber) {
    subscriber = client.duplicate();
    subscriber.on('error', (e: any) => logger.warn('notification.redis', `ws subscriber error: ${String(e?.message).slice(0, 80)}`));
  }
  await subscriber.subscribe(WS_CHANNEL);
  subscriber.on('message', (_channel, raw) => {
    try { handler(JSON.parse(raw) as WsBroadcast); } catch { /* drop malformed */ }
  });
}

export async function closeRedis(): Promise<void> {
  try { await subscriber?.quit(); } catch { /* ignore */ }
  try { await client?.quit(); } catch { /* ignore */ }
  subscriber = null;
  client = null;
}

File: backend/16-recommendation/app/db/redis.py
"""Redis DB 14 — served-feed cache (architecture §3.3).

DEGRADABLE & NON-GATING: a cache miss recomputes from Qdrant/Postgres; a Redis outage
falls straight through to the compute path. Never gates /ready. All ops are best-effort
(return None / no-op on error, never raise).
"""
from __future__ import annotations
import logging
from typing import Optional

from app.config import settings

log = logging.getLogger("reco.redis")

_client = None  # redis.asyncio.Redis | None


async def connect() -> None:
    global _client
    if not settings.redis_host:
        log.info("REDIS_HOST empty — feed cache disabled")
        return
    try:
        import redis.asyncio as aioredis
        _client = aioredis.Redis(
            host=settings.redis_host, port=settings.redis_port,
            password=settings.redis_password or None, db=settings.redis_db,
            socket_connect_timeout=3, socket_timeout=3, decode_responses=True,
        )
        await _client.ping()
        log.info("redis connected db=%d (feed cache)", settings.redis_db)
    except Exception as e:
        log.warning("redis connect deferred (cache degraded): %s", type(e).__name__)
        _client = None


async def health() -> tuple[bool, str]:
    """Diagnostic only — never gates /ready."""
    if _client is None:
        return False, "disabled" if not settings.redis_host else "unreachable"
    try:
        await _client.ping()
        return True, "ok"
    except Exception as e:
        return False, f"err:{type(e).__name__}"


async def get(key: str) -> Optional[str]:
    if _client is None:
        return None
    try:
        return await _client.get(key)
    except Exception:
        return None


async def setex(key: str, value: str, ttl: int) -> None:
    if _client is None:
        return
    try:
        await _client.set(key, value, ex=ttl)
    except Exception:
        pass


async def close() -> None:
    global _client
    if _client is not None:
        try:
            await _client.aclose()
        except Exception:
            pass
    _client = None
